core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -4102,6 +4102,45 @@ public static List reverse(List list) {
     return list;
   }
 
+  /** SQL {@code ARRAY_TO_STRING(array, delimiter)} function. */
+  public static String arrayToString(List list, String delimiter) {
+    return arrayToString(list, delimiter, null);
+  }
+
+  /** SQL {@code ARRAY_TO_STRING(array, delimiter, nullText)} function. */
+  public static String arrayToString(List list, String delimiter, @Nullable String nullText) {
```

Comment:
Maybe `stream` is more simplify.
```
    list.stream()
        .map(item -> {
          if (item == null) {
            return nullText;
          } else if (item instanceof String) {
            return (String) item;
          } else if (item instanceof ByteString) {
            return item.toString();
          } else {
            throw new IllegalStateException(
                "arrayToString supports only String or ByteString, but got "
                    + item.getClass().getName());
          }
        }).collect(Collectors.joining(delimiter));
  }
```

Reply:
Considering performance, I tend to not use lambda. WDYT?

Reply:
Ok